File: android/app/src/main/java/com/fitlog/app/domain/Activity.kt
package com.fitlog.app.domain

/**
 * Actividad importada de otra app: la serie de entrenamientos que traen distancia, duracion y
 * frecuencia cardiaca (Huawei Health o GPX).
 *
 * Es distinta del progreso por ejercicio: una sesion importada no tiene series con peso y reps, asi
 * que sus graficos son de actividad (distancia, tiempo, pulso) y no de volumen.
 */
object Activity {

    enum class Metric(val label: String) {
        DISTANCE("Distancia"),
        DURATION("Duración"),
        HEART_RATE("FC media"),
    }

    data class Input(
        val startedAtMs: Long,
        val finishedAtMs: Long?,
        val distanceM: Double?,
        val averageHeartRate: Double?,
    )

    data class Point(
        val startedAtMs: Long,
        val distanceM: Double,
        val durationMs: Long,
        val averageHeartRate: Double?,
    )

    data class Totals(
        val sessions: Int,
        val distanceM: Double,
        val durationMs: Long,
    )

    /** Serie de actividad dentro del rango, de la mas vieja a la mas nueva. */
    fun series(inputs: List<Input>, range: Progress.Range? = null): List<Point> =
        inputs
            .filter { input -> inRange(input.startedAtMs, range) }
            .sortedBy { it.startedAtMs }
            .map { input ->
                Point(
                    startedAtMs = input.startedAtMs,
                    distanceM = input.distanceM ?: 0.0,
                    durationMs = input.finishedAtMs
                        ?.let { (it - input.startedAtMs).coerceAtLeast(0L) }
                        ?: 0L,
                    averageHeartRate = input.averageHeartRate,
                )
            }

    /** Valor de la metrica elegida para un punto, o `null` si el entrenamiento no la trae. */
    fun value(point: Point, metric: Metric): Double? = when (metric) {
        Metric.DISTANCE -> point.distanceM.takeIf { it > 0 }
        Metric.DURATION -> point.durationMs.takeIf { it > 0 }?.toDouble()
        Metric.HEART_RATE -> point.averageHeartRate
    }

    /** Puntos con valor para la metrica elegida: los que no la traen no se grafican. */
    fun values(points: List<Point>, metric: Metric): List<Double> =
        points.mapNotNull { value(it, metric) }

    fun totals(points: List<Point>): Totals = Totals(
        sessions = points.size,
        distanceM = points.sumOf { it.distanceM },
        durationMs = points.sumOf { it.durationMs },
    )

    private fun inRange(timestampMs: Long, range: Progress.Range?): Boolean {
        if (range == null) return true
        val from = range.fromMs
        if (from != null && timestampMs < from) return false
        return timestampMs <= range.toMs
    }
}

File: web/src/domain/activity.ts
import type { ProgressRange } from '@/domain/progress';

/**
 * Actividad importada de otra app: la serie de entrenamientos que traen distancia, duración y
 * frecuencia cardíaca (Huawei Health o GPX).
 *
 * Es distinta del progreso por ejercicio: una sesión importada no tiene series con peso y reps, así
 * que sus gráficos son de actividad (distancia, tiempo, pulso) y no de volumen.
 */

export const ACTIVITY_METRICS = ['distance', 'duration', 'heart_rate'] as const;

export type ActivityMetric = (typeof ACTIVITY_METRICS)[number];

export const ACTIVITY_METRIC_LABELS: Record<ActivityMetric, string> = {
  distance: 'Distancia',
  duration: 'Duración',
  heart_rate: 'FC media',
};

export interface ActivityInput {
  readonly startedAtMs: number;
  readonly finishedAtMs: number | null;
  readonly distanceM: number | null;
  readonly averageHeartRate: number | null;
}

export interface ActivityPoint {
  readonly startedAtMs: number;
  readonly distanceM: number;
  readonly durationMs: number;
  readonly averageHeartRate: number | null;
}

export interface ActivityTotals {
  readonly sessions: number;
  readonly distanceM: number;
  readonly durationMs: number;
}

/** Serie de actividad dentro del rango, de la más vieja a la más nueva. */
export function buildActivitySeries(
  inputs: readonly ActivityInput[],
  range: ProgressRange | null = null
): ActivityPoint[] {
  return inputs
    .filter((input) => inRange(input.startedAtMs, range))
    .slice()
    .sort((a, b) => a.startedAtMs - b.startedAtMs)
    .map((input) => ({
      startedAtMs: input.startedAtMs,
      distanceM: input.distanceM ?? 0,
      durationMs:
        input.finishedAtMs === null
          ? 0
          : Math.max(0, input.finishedAtMs - input.startedAtMs),
      averageHeartRate: input.averageHeartRate,
    }));
}

/** Valor de la métrica elegida para un punto, o `null` si el entrenamiento no la trae. */
export function activityValue(point: ActivityPoint, metric: ActivityMetric): number | null {
  if (metric === 'distance') return point.distanceM > 0 ? point.distanceM : null;
  if (metric === 'duration') return point.durationMs > 0 ? point.durationMs : null;
  return point.averageHeartRate;
}

/** Puntos con valor para la métrica elegida: los que no la traen no se grafican. */
export function activityValues(
  points: readonly ActivityPoint[],
  metric: ActivityMetric
): number[] {
  return points
    .map((point) => activityValue(point, metric))
    .filter((value): value is number => value !== null);
}

export function activityTotals(points: readonly ActivityPoint[]): ActivityTotals {
  return {
    sessions: points.length,
    distanceM: points.reduce((total, point) => total + point.distanceM, 0),
    durationMs: points.reduce((total, point) => total + point.durationMs, 0),
  };
}

function inRange(timestampMs: number, range: ProgressRange | null): boolean {
  if (range === null) return true;
  if (range.fromMs !== null && timestampMs < range.fromMs) return false;
  return timestampMs <= range.toMs;
}
